File: ICPC/WarmUp 10/ncpc2015solutions/bells_lukas.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Solution by lukasP (Lukáš Poláček)
from sys import stdin

def gen(n):
    if n == 1:
        yield [1]
        return

    dir = False
    for enumer in gen(n - 1):
        o = list(enumer)
        if not dir:
            o.append(n)
        else:
            o.insert(0, n)
        yield o

        cur = 0 if dir else n - 1
        for i in range(n - 1):
            pos = cur + (1 if dir else -1)
            newo = list(o)

            newo[cur], newo[pos] = newo[pos], newo[cur]
            yield newo

            cur = pos
            o = newo
        dir = not dir

n = int(stdin.readline())
for row in gen(n):
    print(" ".join(map(str, row)))

File: ICPC/WarmUp 10/ncpc2015solutions/bells_lukas.cs
//Solution by lukasP (Lukáš Poláček)
using System;
using System.Linq;
using System.Collections.Generic;

class Problem
{
    public static IEnumerable<IEnumerable<int>> gen(int n) {
        if (n == 1)
        {
            yield return Enumerable.Repeat(1, 1);
            yield break;
        }

        bool dir = false;
        foreach (IEnumerable<int> enumer in gen(n - 1)) {
            List<int> o = enumer.ToList();
            if (!dir) o.Add(n);
            else o.Insert(0, n);
            yield return o;

            int cur = dir ? 0 : n - 1;
            for (int i = 0; i < n - 1; i++) {
                int pos = cur + (dir ? 1 : -1);
                List<int> newo = new List<int> (o);

                int tmp = newo[cur];
                newo[cur] = newo[pos];
                newo[pos] = tmp;
                yield return newo;

                cur = pos;
                o = newo;
            }
            dir = !dir;
        }
    }

    public static void Main ()
    {
        int n = int.Parse(Console.ReadLine());
        foreach (IEnumerable<int> row in gen(n))
            Console.WriteLine(String.Join(" ", row));
    }
}
